flang/lib/Optimizer/CodeGen/CodeGen.cpp

```diff
@@ -2424,11 +2438,53 @@ struct GlobalOpConversion : public FIROpConversion<fir::GlobalOp> {
     auto g = rewriter.create<mlir::LLVM::GlobalOp>(
         loc, tyAttr, isConst, linkage, global.sym_name(), initAttr);
     auto &gr = g.getInitializerRegion();
+    if (global.getType().isa<fir::SequenceType>() &&
+        !global.getType().dyn_cast<fir::SequenceType>().hasUnknownShape() &&
+        global.hasInitializationBody()) {
+      if (auto singleInsertOp = getSingleOpOfType<fir::InsertOnRangeOp>(
+              global.region().front())) {
+        auto globalType = global.getType().dyn_cast<fir::SequenceType>();
+        // Check if the initialization is on the full range.
+        if (getNumElementInRange(singleInsertOp.coor()) ==
+            globalType.getNumElements()) {
+          rewriter.inlineRegionBefore(global.region(), gr, gr.end());
+          auto insertOp = getSingleOpOfType<fir::InsertOnRangeOp>(gr.front());
+          auto op = insertOp.val().getDefiningOp();
+          auto constant = dyn_cast<mlir::ConstantOp>(op);
+          if (!constant) {
+            auto convertOp = dyn_cast<fir::ConvertOp>(op);
+            constant =
+                dyn_cast<mlir::ConstantOp>(convertOp.value().getDefiningOp());
+          }
+          mlir::Type vecType =
+              mlir::VectorType::get(globalType.getShape(), constant.getType());
+          auto denseAttr = mlir::DenseElementsAttr::get(
+              vecType.cast<ShapedType>(), constant.getValue());
+          rewriter.setInsertionPointAfter(insertOp);
+          rewriter.replaceOpWithNewOp<mlir::ConstantOp>(insertOp, tyAttr,
+                                                        denseAttr);
+          rewriter.eraseOp(global);
+          g.typeAttr(mlir::TypeAttr::get(tyAttr));
+          return success();
+        }
+      }
+    }
     rewriter.inlineRegionBefore(global.region(), gr, gr.end());
     rewriter.eraseOp(global);
     return success();
   }
 
+  unsigned getNumElementInRange(mlir::ArrayAttr indexes) const {
```

Comment:
insert_on_range is a bit unusual, it is not describing a "square" section like in Fortran, it's including all the points belonging to the array element sequence between two points (see discussion [here](https://github.com/flang-compiler/f18-llvm-project/pull/1050#discussion_r706518308))

So to compute the number of elements between the points, you need the extent, end the formula is:

`n = (y0-x0)+(y1-x1)*exent0 + .... + (yn-xn)*exent(n-1)` and not `n = (y0-x0)*...*(y1-x1)`

They  match when (x0, ..., xn) is the first element and (y0, ..., yn) is the last element, but it's otherwise not the same, and with the potential underflow from yi-xi that can be negative, I am not sure false match are not possible.

If all you need is to check is that you have the first and last point, it might be easier to check that the first is (0, .... ,0), and the last is (extent0-1, ..., extentn-1).

Reply:
Ok thanks for the precision here. I think we just need to make sure we are working on the full range so checking that we have 0 and extent _n_ -1 should do the trick.